Streamline Task Types

jdolence opened the pull request:
<!--Provide a general summary of your changes in the title above, for
example "Add AMR unit test for cell centered fields.".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

## PR Summary

One of the issues exposed in #189 related to the fact that `Container`s were being passed by value into tasks, which led to `VariablePack`s being reconstructed every cycle.  This was addressed in #196 by storing pointers to the `Container`s in the task, but this suggested to me that our model where developers define a functor for every type of task was clumsy and error prone.  This PR redesigns the tasks into a single `Task` type to capture generic task signatures.  Arguments are _always_ captured by value now via a lambda, so developers need to design to that.  To address the `VariablePack` construction issue and to avoid the expense of copying `Container`s by value, the `ContainerCollection` now returns `std::shared_ptr<Container<T>>` and this is the object passed around to tasks that formerly took `Container<T>` directly. 

<!--Please provide at least 1-2 sentences describing the pull request in
detail.  Why is this change required?  What problem does it solve?-->

<!--If it fixes an open issue, please link to the issue here.-->

## PR Checklist

<!-- Note that some of these check boxes may not apply to all pull requests -->

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Code is formatted

AndrewGaspar commented:
> To address the VariablePack construction issue, member data in the Container object are now std::shared_ptrs which are trivially copied as necessary. 

`std::shared_ptr` technically isn't "trivially" copyable, which [has a specific definition in C++](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable).

Wouldn't it actually make more sense to instead wrap `Container` in `std::shared_ptr`? As far as I can tell, you never hand out copies of the internal shared pointers - it's only at the level of Container that they are copied, right? That would make the copy quite a bit cheaper, and I think would be easier to reason about (i.e. the entire structure is shared rather than just the maps inside it).

jdolence commented:
> > To address the VariablePack construction issue, member data in the Container object are now std::shared_ptrs which are trivially copied as necessary.
> 
> `std::shared_ptr` technically isn't "trivially" copyable, which [has a specific definition in C++](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable).
> 
> Wouldn't it actually make more sense to instead wrap `Container` in `std::shared_ptr`? As far as I can tell, you never hand out copies of the internal shared pointers - it's only at the level of Container that they are copied, right? That would make the copy quite a bit cheaper, and I think would be easier to reason about (i.e. the entire structure is shared rather than just the maps inside it).

Yes probably.  Sorry for the imprecise language -- just meant it was comparatively cheap (relative to deep copies of maps and vectors and such).  This change was transparent to the rest of the code, and allows us to continue passing around references to `Container<T>`.  With your suggestion, we'd have to move to passing around `std::shared_ptr<Container<T>>`, which is fine.  I'll take a look and might give this a go.

Thanks for the suggestion!

AndrewGaspar commented:
> With your suggestion, we'd have to move to passing around `std::shared_ptr<Container<T>>`, which is fine. I'll take a look and might give this a go.

Something you could do is make the `shared_ptr` internal to `Container`. For example:

Before:
```c++
class Container {
public:
  int GetMember1() const { return member1_; }
  float GetMember2() const { return member2_; }
private:
  int member1_;
  float member2_;
};
```

After:
```c++
class Container {
public:
  int GetMember1() const { return state_->member1_; }
  float GetMember2() const { return state_->member2_; }
private:
  struct State {
    int member1_;
    float member2_;
  };

  std::shared_ptr<State> state_;
};
```

jdolence commented:
@AndrewGaspar went ahead and moved to `std::shared_ptr<Container<T>>`  I agree with your original point that this makes things a bit easier to reason about -- we know how std::shared_ptr behaves upon copying, but you'd have to look in the guts of `container.hpp` or find something in the docs to understand how `Container<T>` copies.  Let me know if you've got more thoughts on this.

Also replaced the `std::bind` with a lambda.  Thoughts?

AndrewGaspar commented:
My opinion is to always prefer a lambda to `std::bind`.

AndrewGaspar commented on example/face_fields/face_fields_example.hpp:
```diff
@@ -20,7 +20,7 @@
 #include "interface/state_descriptor.hpp"
 #include "mesh/mesh.hpp"
-#include "task_list/tasks.hpp"
+//#include "task_list/tasks.hpp"
```
```suggestion
```

jdolence commented on example/advection/advection_driver.cpp:
```diff
@@ -49,10 +50,10 @@ TaskStatus UpdateContainer(MeshBlock *pmb, int stage,
-  Container<Real> &cin = pmb->real_containers.Get(stage_name[stage - 1]);
-  Container<Real> &cout = pmb->real_containers.Get(stage_name[stage]);
-  Container<Real> &dudt = pmb->real_containers.Get("dUdt");
+  auto &base = pmb->real_containers.Get();
+  auto &cin = pmb->real_containers.Get(stage_name[stage - 1]);
+  auto &cout = pmb->real_containers.Get(stage_name[stage]);
+  auto &dudt = pmb->real_containers.Get("dUdt");
```
Yeah I don't know what best practices might be here.  To me, `std::shared_ptr` is just another object and I'd rather not just go around copying objects needlessly.  That may well be a flawed argument.  Also, it's possible these should really be something more like `const auto &` but what do I know.

Yurlungur replied:
Yeah my logic is just that a shared pointer is just a reference-counted pointer. And a reference is just a pointer you don't have to de-reference. So this is like a pointer of a pointer. But your point is well taken. The reference semantics reduce copies of whatever is going on under the hood in shared pointer. I don't think it matters much. If you can add const, I like that.

jdolence replied:
yeah, ok.  not going to add `const` for this one, but it might be good to try to remember in the future.

AndrewGaspar replied:
Unnecessary copies of shared pointers has been shown to have performance implications in certain cases. I don't think it applies here.

Basically, a temporary copy of a shared pointer results in an atomic increment on copy and decrement when it goes out of scope. So, it can result in some unnecessary cache line invalidation. Probably doesn't matter much here.

Yurlungur replied:
Ah. Cool. Thanks for the explanation, @AndrewGaspar !

jdolence commented on src/tasks/task_list.hpp:
```diff
@@ -0,0 +1,115 @@
+    return TaskListStatus::running;
+  }
+
+  template <class F, class... Args>
+  TaskID AddTask(F func, TaskID &dep, Args &&... args) {
+    TaskID id(tasks_added_ + 1);
+    task_list_.push_back(
+        Task(id, dep, [=]() mutable -> TaskStatus { return func(args...); }));
+    tasks_added_++;
+    return id;
+  }
```
thanks!

jdolence commented:
> This update sure streamlines the tasks, and seems to make them more general. Until I try using this new task design in our own multi-physics code, I am not sure I understand all the implications of these changes. However, it certainly looks good to me and I approve merging. Once I try them out I might have more feedback.
> I note a much older change seems to be the switch to C++ std::bitset class, rather than manipulating raw bits in int or long int. The latter was done for performance (not that we ever saw the Task Manager slowing down the code), but I just want to be sure the chance to STL hasn't affected anything adversely. I admit it gets rid of custom code to check bits, which makes the source cleaner.

Thanks for taking a look @jmstone!  I'll go ahead and merge -- we can always revisit if needed.

The change to std::bitset happened a while ago, as you said.  It was motivated by the fact that uint64_t (or similar) only supports a fixed number of tasks (bits), and I worried that down the road we'd end up with very difficult to identify bugs.  Such bugs would have been easier to identify in the original design because tasks were explicitly numbered, but that's no longer the case.  We can certainly revisit this if it becomes a performance issue, but I suspect it won't.  On this topic, I'm not sure my particular implementation -- using a std::vector<std::bitset<SOME_BLOCK_SIZE>> -- is ideal, but it works so it's probably not worth worrying about until (and if) it becomes an issue.  It does require some custom code, though, as you may have already noticed in task_id.cpp.
